src/queries.ts

```diff
@@ -26,44 +27,63 @@ interface AndroidQueryParams extends BaseQueryParams {
   bid_browsers?: string[];
 }
 
+function isDesktopParams(object: any): object is DesktopQueryParams {
+  return 'bid_window' in object;
+}
+
+function isAndroidParams(object: any): object is AndroidQueryParams {
+  return 'bid_android' in object;
+}
+
 // Constructs a query that returns a fully-detailed list of events from the merging of several sources (window, afk, web).
-// Puts it's results in `not_afk` and `events`.
-function canonicalEvents(params: DesktopQueryParams): string {
+// Performs:
+//  - AFK filtering (if filter_afk is true)
+//  - Categorization (if classes specified)
+//  - Filters by category (if filter_classes set)
+// Puts it's results in `events` and `not_afk` (if not_afk available for platform).
+function canonicalEvents(params: DesktopQueryParams | AndroidQueryParams): string {
+  // Needs escaping for regex patterns like '\w' to work (JSON.stringify adds extra unecessary escaping)
+  const classes_str = JSON.stringify(params.classes).replace('\\\\', '\\');
+  const cat_filter_str = JSON.stringify(params.filter_classes);
+  const bid_window = isDesktopParams(params) ? params.bid_window : params.bid_android;
+
   return `
-    events  = flood(query_bucket("${params.bid_window}"));
-    not_afk = flood(query_bucket("${params.bid_afk}"));
-    not_afk = filter_keyvals(not_afk, "status", ["not-afk"]);
-    ${params.filter_afk ? 'events  = filter_period_intersect(events, not_afk);' : ''}
+    events = flood(query_bucket("${bid_window}"));
+    ${
+      isDesktopParams(params)
+        ? `not_afk = flood(query_bucket("${params.bid_afk}"));
+           not_afk = filter_keyvals(not_afk, "status", ["not-afk"]);`
+        : ''
+    }
+    ${isAndroidParams(params) ? 'events = merge_events_by_keys(events, ["app"]);' : ''}
```

Comment:
What's going on here, only merge them for android?
How does that make any sense, the events will be completely different on android and desktop?
If you do that here the I wouldn't really call this function "canonical" as the output is completely different.
Seems like it is merged anyway in appQuery, but differently for title and appname? (which looks correct, while this does not).

Reply:
They are merged here to avoid potentially returning several thousand events per day, same as happened in the android query before.

>If you do that here the I wouldn't really call this function "canonical" as the output is completely different.

If you want to call it something else then I'm open to suggestions.

>Seems like it is merged anyway in appQuery, but differently for title and appname? (which looks correct, while this does not).

It worked the same way before, I just moved the line.